File: frontend/src/services/savedQueryService.ts
import { httpDelete, httpGet, httpPost, httpPut } from "@/services/httpClient";

export interface SavedQuery {
  id: number;
  user_id: number | null;
  name: string;
  description: string | null;
  sql: string;
  database: string;
  is_favorite: boolean;
  created_at: string;
  updated_at: string;
}

export interface SavedQueryCreate {
  name: string;
  description?: string;
  sql: string;
  database?: string;
  is_favorite?: boolean;
}

export interface SavedQueryUpdate {
  name?: string;
  description?: string;
  sql?: string;
  database?: string;
  is_favorite?: boolean;
}

export interface SavedQueryListResponse {
  items: SavedQuery[];
}

export const savedQueryService = {
  create: async (data: SavedQueryCreate): Promise<SavedQuery> => {
    return httpPost<SavedQuery>("/v1/saved-queries", data);
  },

  list: async (favoritesOnly: boolean = false): Promise<SavedQueryListResponse> => {
    const qs = favoritesOnly ? "?favorites_only=true" : "";
    return httpGet<SavedQueryListResponse>(`/v1/saved-queries${qs}`);
  },

  get: async (queryId: number): Promise<SavedQuery> => {
    return httpGet<SavedQuery>(`/v1/saved-queries/${queryId}`);
  },

  update: async (queryId: number, data: SavedQueryUpdate): Promise<SavedQuery> => {
    return httpPut<SavedQuery>(`/v1/saved-queries/${queryId}`, data);
  },

  delete: async (queryId: number): Promise<void> => {
    await httpDelete(`/v1/saved-queries/${queryId}`);
  }
};

File: backend/app/services/saved_query_service.py
"""SavedQuery Service"""

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.saved_query import SavedQuery
from app.repositories.saved_query_repository import saved_query_repository
from app.schemas.saved_query_schema import SavedQueryCreate, SavedQueryUpdate


class SavedQueryService:
    """저장된 쿼리 서비스"""

    async def create_query(
        self, db: AsyncSession, data: SavedQueryCreate, user_id: int | None = None
    ) -> SavedQuery:
        """쿼리 저장"""
        return await saved_query_repository.create(
            db,
            user_id=user_id,
            name=data.name,
            description=data.description,
            sql=data.sql,
            database=data.database,
            is_favorite=data.is_favorite,
        )

    async def get_query(self, db: AsyncSession, query_id: int) -> SavedQuery | None:
        """쿼리 조회"""
        return await saved_query_repository.get_by_id(db, query_id)

    async def list_queries(
        self, db: AsyncSession, user_id: int | None = None, favorites_only: bool = False
    ) -> list[SavedQuery]:
        """쿼리 목록 조회"""
        return await saved_query_repository.get_all(db, user_id, favorites_only)

    async def update_query(
        self, db: AsyncSession, query_id: int, data: SavedQueryUpdate
    ) -> SavedQuery | None:
        """쿼리 수정"""
        query = await saved_query_repository.get_by_id(db, query_id)
        if not query:
            return None

        return await saved_query_repository.update(
            db,
            query,
            name=data.name,
            description=data.description,
            sql=data.sql,
            database=data.database,
            is_favorite=data.is_favorite,
        )

    async def delete_query(self, db: AsyncSession, query_id: int) -> bool:
        """쿼리 삭제"""
        query = await saved_query_repository.get_by_id(db, query_id)
        if not query:
            return False

        await saved_query_repository.delete(db, query)
        return True


saved_query_service = SavedQueryService()
